File: Algorithms-DataStructures-in-C/BinaryHeap.c
//#include <stdio.h>
//
//typedef int boolean;
//#define false 0
//#define true 1
//
//int pq[16];
//int N = 0;
//
//boolean less(int i, int j) {
//	return pq[i] < pq[j];
//}
//
//void exchange(int i, int j) {
//	int temp = pq[i];
//	pq[i] = pq[j];
//	pq[j] = temp;
//}
//
//void swim(int k) {
//	while (k > 1 && less(k/2,k)) {
//		exchange(k, k / 2);
//		k = k / 2;
//	}
//}
//
//void sink(int k) {
//
//	while (2 * k <= N) {
//		int j = 2 * k;
//		if (j < N && less(j, j + 1))
//			j = j + 1;
//
//		if (less(k, j)) {
//			exchange(k, j);
//			k = j;
//		} else
//			break;
//	}
//}
//
//void insert(int x) {
//	pq[++N] = x;
//	swim(N);
//}
//
//int deletemax() {
//	int max = pq[1];
//	exchange(1, N--);
//	sink(1);
//	return max;
//}
//
//int main(int argc, char **argv) {
//
//	insert(25);
//	insert(10);
//	insert(5);
//	insert(15);
//	insert(30);
//	insert(20);
//
//	printf("%d\n", deletemax());
//	printf("%d\n", deletemax());
//	printf("%d\n", deletemax());
//	printf("%d\n", deletemax());
//	printf("%d\n", deletemax());
//	printf("%d\n", deletemax());
//	return 0;
//}

File: Algorithms-DataStructures-in-Java/BinaryHeap.java
//Program to implement Heap

public class BinaryHeap<Key extends Comparable<Key>> {
	private Key pq[];
	private int N;

	public BinaryHeap(int capacity) {
		pq = (Key[]) new Comparable[capacity + 1];
	}

	public void sink(int k) {
		while (2 * k <= N) {
			int j = 2 * k;
			if (j < N && less(j, j + 1))
				j = j + 1;
			if (less(k, j)) {
				exch(k, j);
				k = j;
			} else
				break;
		}
	}

	public void swim(int k) {

		while (k > 1 && less(k / 2, k)) {
			exch(k, k / 2);
			k = k / 2;
		}
	}

	public void insert(Key x) {
		pq[++N] = x;
		swim(N);
	}

	public Key deleteMax() {
		Key max = pq[1];
		exch(1, N--);
		sink(1);
		return max;
	}

	public boolean isEmpty() {
		return N == 0;
	}

	public boolean less(int v, int w) {
		return pq[v].compareTo(pq[w]) < 0;
	}

	public void exch(int i, int j) {
		Key temp = pq[i];
		pq[i] = pq[j];
		pq[j] = temp;
	}

	public static void main(String[] args) {
		BinaryHeap<Integer> obj = new BinaryHeap<>(10);
		obj.insert(25);
		obj.insert(20);
		obj.insert(05);
		obj.insert(15);
		obj.insert(10);
		obj.insert(30);
		System.out.println(obj.deleteMax());
		System.out.println(obj.deleteMax());
		System.out.println(obj.deleteMax());
		System.out.println(obj.deleteMax());
		System.out.println(obj.deleteMax());
		System.out.println(obj.deleteMax());
	}

}
